harvester/harvester
# WIP: Add volumesnapshotnameclass to vmbackup

**Problem:**


**Solution:**


**Related Issue:**
#553 

**Test plan:**

Setup:
* Create a harvester cluster from master-head.
* Create a VM `source-vm`.
* Run command `echo "123" > test.txt && sync` in `source-vm`.

Case 1: New fields don't break old VMBackup
* Configure Backup Target.
* Create a VMBackup `vm-backup-old`.
* Clone this branch and run following commands under `harvester/harvester/deploy/charts`:
```
> helm upgrade harvester-crd harvester-crd --namespace harvester-system
> helm upgrade harvester harvester --namespace harvester-system --set-string containers.apiserver.image.repository=frankyang/harvester --set-string containers.apiserver.image.tag=add-volumesnapshotnameclass-to-vmbackup-head --set-string webhook.image.repository=frankyang/harvester-webhook --set-string webhook.image.tag=add-volumesnapshotnameclass-to-vmbackup-head
```
* After upgrade done, check `vm-backup-old.spec.type` is `backup`, `vm-backup-old.status.csiDriverVolumeSnapshotClassNameMap` is not empty, and eacho `vm-backup-old.status.volumeBackup[]` has `csiDriverName`.
* Restore `vm-backup-old` to a new VM and replace `source-vm`. Both of VMs should have `test.txt` file with content `123`.
* Cleanup VMs and VMBackup.

Case 2: Create a snapshot for `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineBackup
metadata:
  name: vm-snapshot
  namespace: default
spec:
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  type: "snapshot"
```
* VMBackup `vm-snapshot` should be created successfully.

Case 3: Restore `vm-snapshot` to a new VM `vm-from-snapshot`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: restore-vm-from-snapshot
  namespace: default
spec:
  newVM: true
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-from-snapshot
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
```
* A new VM `vm-from-snapshot` should be created successfully.
* Run command `cat test.txt` in `vm-from-snapshot` and it should show content `123`.

Case 4: Replace `source-vm` with `vm-snapshot`.
* Run command `echo "456" > test.txt && sync` in `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: replace-vm-from-snapshot
  namespace: default
spec:
  newVM: false
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  virtualMachineBackupName: vm-snapshot
  virtualMachineBackupNamespace: default
```
* VM `source-vm` should be restored successfully.
* Run command `cat test.txt` in `source-vm` and it should show content `123`.

Case 5: Create a backup for `source-vm`.
* Setup backup target.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineBackup
metadata:
  name: vm-backup
  namespace: default
spec:
  source:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  type: "backup"
```
* VMBackup `vm-backup` should be created successfully.

Case 6: Restore `vm-backup` to a new VM `vm-from-backup`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: restore-vm-from-backup
  namespace: default
spec:
  newVM: true
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: vm-from-backup
  virtualMachineBackupName: vm-backup
  virtualMachineBackupNamespace: default
```
* A new VM `vm-from-backup` should be created successfully.
* Run command `cat test.txt` in `vm-from-snapshot` and it should show content `123`.

Case 7: Replace `source-vm` with `vm-backup`.
* Run command `echo "456" > test.txt && sync` in `source-vm`.
* Apply the following YAML.
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineRestore
metadata:
  name: replace-vm-from-snapshot
  namespace: default
spec:
  newVM: false
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: source-vm
  virtualMachineBackupName: vm-backup
  virtualMachineBackupNamespace: default
```
* VM `source-vm` should be restored successfully.
* Run command `cat test.txt` in `source-vm` and it should show content `123`.
